### ISSUE TITLE
How to update/query/scan in DynamoDB using Amplify?

### ISSUE THREAD

**Issue by chai86** [aws-amplify/amplify-js#2463]
I would like to update a field in my DynamoDB table called 'AppTable' using the AWS-Amplify methods that were generated when I created the corresponding CRUD cloud-api (e.g. PUT, GET, POST, DELETE).

These few API methods reside in awsmobilejs/backend/cloud-api/AppTable/app.js, but they are not comprehensive enough!

I can simply use API.put to add to this table. e.g:

`apiResponse = API.put('AppTableCRUD', path, body)`
```

body: {
    "uploaderBool": true,
    "userId": 'user1',
    "itemId": '10005',
    "Username": "first_app_user",
    "Email": "user@myapp.com",
    "NumberList": 
       [
        "7.9",
        "5.7",
        "3.4",
        "4.9"
       ],
    "AverageNumber":[
        "5.5"
       ],
     }
    }

```
What i want to do, is update the NumberList field with a function in my react-native app, count how many numbers in the list on DynamoDB, then calculate/write an updated number called AverageNumber.

If i use a current API.put method i have to put in this ENTIRE body everytime (this list could be hundreds of thousand of entries so Get, then Put with a single update is absurd).

How do you use the **.scan()**, **.query()** and **.update()** method that DynamoDB has with Amplify?

**Comment by elorzafe** [comment]
@chai86 

I tried this on a similar project with a code like this.

I modified my `app.js` file that contains the [aws serverlesss express code](https://github.com/awslabs/aws-serverless-express) that runs on my lambda function by adding support for PATCH method. This has a path `/items/addNumber/:hashKey/:sortKey` where :hashKey and :sortKey are the object that is going to be updated. Inside that function I get the data from DynamoDB (including the array of numbers), then adding the new number to the array and calculating the average. `newNumber` is on the body of the request.

I added this code on app.js (lambda function code)
```javascript
app.patch(path + "/addNumber" + hashKeyPath + sortKeyPath, function (req, res) {
  let params = {}
  try {
    params[partitionKeyName] = convertUrlType(req.params[partitionKeyName], partitionKeyType);
  } catch (err) {
    res.json({ error: 'Wrong column type ' + err });
  }

  if (hasSortKey) {
    try {
      params[sortKeyName] = convertUrlType(req.params[sortKeyName], sortKeyType);
    } catch (err) {
      res.json({ error: 'Wrong column type ' + err });
    }
  }

  let getItemParams = {
    TableName: tableName,
    Key: params
  }

  var newNumber = req.body.newNumber;

  if (newNumber == null || isNaN(newNumber)) {
    res.json({ error: 'Wrong number ' + newNumber });
  }

  dynamodb.get(getItemParams, (err, data) => {
    if (err) {
      res.json({ error: 'Could not load items: ' + err.message });
    } else {
      if (data.Item) {
        data = data.Item;
      }
      data.numbers = [].concat(data.numbers || []).concat(newNumber);
      
      data.average = data.numbers.reduce(function(acc, currentValue){ return acc + currentValue; }) / data.numbers.length;

      let putItemParams = {
        TableName: tableName,
        Item: data
      }
      dynamodb.put(putItemParams, (err, data) => {

        if (err) {
          res.json({ error: err, url: req.url, body: req.body });
        } else {
          res.json({ success: 'post call succeed!', url: req.url, data: data })
        }
      });
    }
  });
```

After updating the code I did a `push` to update the lambda code that runs on the server.

Now my client code to invoke the api is like this.  

```javascript
API.patch('myapi', '/items/addNumber/id-1/item-1', {
      body: {
        "newNumber": 3
      }
    }).then(result => console.log({result}));
```

`id-1` is a value from my hashKey and `item-1` the sortKey. The body of the request contains the `newNumber`.

I hope this helps

**Comment by chai86** [comment]
@elorzafe 
Ah ok. I'm still unsure why if the `req.body` i put in as below, it doesn't work. 

   ```
var updateTableWith = {
              TableName: 'AppTable',
              Key:{
                'userId': 'user1',
                'itemId': '10005'
              },
              UpdateExpression: 'set NumberList= :r', AverageNumber= :f',
              ExpressionAttributeValues: { 
                ":r": ["7.3", "0.7", "9.9"],
                ':f': "10.0"
            },
            ReturnValues: "UPDATED_NEW"        
        }
```
In the documentation for DynamoDB the `params` i thought the should `updateTableWith` should still get processed when calling update() as that's what the function is expecting right?
```
var docClient = new AWS.DynamoDB.DocumentClient();
docClient.update(params, function(err, data)
```

AWS AppSync with mutations, subscriptions etc... seems quite counter intuitive when DynamoDB's Put, Get, Delete, Scan, Query, Update functions etc.... are pretty self explanatory and directly address my DynamoDB database. Plus AppSync is schema based.

Nevertheless, i have spent months configuring a React Native mobile app, uploading images, videos, designing a front end etc... assuming the database piece would be relatively straightforward. I weighed up the options and chose to utilise AWS Amplify and AWS MobileHub to develop my mobile app, and for Amplify to not offer some of the fundamental database functions in the `backend/cloud-api/AppTable/app.js` is a huge surprise at this stage. You guys have done fantastic work on the `aws-mobile-react-native-starter`, truly helpful and avant-garde!

Right now when I type `app.xxxx()` in `backend/cloud-api/AppTable/app.js` the functions dont appear. Will the `dynamodb.query()`, `dynamodb.scan()` and `dynamodb.update()` ever be made available @powerful23 @mlabieniec @mbahar @richardzcode ?

**Comment by elorzafe** [comment]
@chai86 

Correct me if I didn't understand well.

What you are saying is you would like to have something like this on your app code.

`const apiResponse = await API.scan('myappAPI', path, options)` and that `scan` will be on your lambda function code. So it will be a 1:1 parity between the app code and the backend code and all DynamoDB methods will be available?

Thanks for your feedback!

**Comment by chai86** [comment]
@elorzafe 

Thats spot on! There's no use to trying to use AWS mobile with DynamoDB with only a put, get and delete. I've been left "paralyzed" in project terms for 2 weeks now because of the absence of the other functions. 

How can I achieve this?: 

```
const apiResponse = await API.scan('myappAPI', path, options)
const apiResponse = await API.update('myappAPI', path, options)
const apiResponse = await API.query('myappAPI', path, options)
```

p.s. also i guess there would need to be examples of what the 'options' paramater would contain and how it would be composed.

**Comment by elorzafe** [comment]
@chai86 

Is an interesting use case what you mention (I am sorry you have been paralyzed in your project), but API category was designed to make HTTP requests to REST and Graphql endpoints, is more than the specific use case as a data proxy to DynamoDB. 

We provide capabilitites to integrate with DynamoDB by generating lambda function code that supports this. I understand that for your requirements that is not enough, that's why I recommend to you to use AppSync (https://aws-amplify.github.io/docs/js/api#simple-query) in my opinion is more simple but is ok if you don't think the same. I can discuss this with the team so I will mark this issue as feature request.

**Comment by elorzafe** [comment]
@chai86, do you have a annotated schema? (Assuming you used amplify-cli). I can help you if you want?

**Comment by chai86** [comment]
Thanks @elorzafe 

My schema.graphql is really simple:

```
type Users @model {
  id: ID!
  userName: String!
  name: String
  email: String!
  mobileNum: String
  listOfRated: [RatedPairs]
  userDescription: String
  isBlocked: Boolean
}

type Video @model {
  id: ID!
  videoIdInt: Int
  videoName: String!
  videoUploader: String!
  videoUploadedAt: String
  videoLength: Float!
  numberList: [Float]!
  shareLink: String
}

type RatedPairs {
  videoId: Int
  correspondingRating: Float
}
```

In the example in the weblink i previously sent, it only attempts a write to the DynamoDB video table using my createVideo mutation (again in the weblink)

**Comment by elorzafe** [comment]
Hi @chai86, I created graphql api using aws-amplify-cli with this app code (react web)

```jsx
import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import Amplify, { API, graphqlOperation } from 'aws-amplify';
import { withAuthenticator } from "aws-amplify-react";
import config from './aws-exports';
import { createVideo } from "./graphql/mutations";

Amplify.configure(config);

class App extends Component {

  CreateVideo = async () => {

    const newVideoInfo = {
      videoIdInt: 12345,
      videoName: "sheeran-moon",
      videoUploader: "ed",
      videoUploadedAt: "tonight",
      videoLength: 85,
      numberList: [7, 8, 9],
      shareLink: "www.voda.com"
    }

    try {
      await API.graphql(graphqlOperation(createVideo, { input: newVideoInfo }))
      console.log('User successfully created!!')
    }
    catch (err) {
      console.log('create user error: ', err)
    }

  }

  render() {
    return (
      <div >
        <button
          onClick={this.CreateVideo}
        >Add user</button>
      </div>

    );
  }


}

export default withAuthenticator(App, true);
```

it  seems that `input` attribute is missing so it should be like this
`await API.graphql(graphqlOperation(createVideo, { input: newVideoInfo }))`

**Comment by chai86** [comment]
@elorzafe 

You've only gone and bloomin done it! Thanks!

Right, now to play with the other mutations and queries (create, delete, update, get, list). I'll let you know how i go 👍

**Comment by elorzafe** [comment]
@chai86 I am glad it works! 😄

**Comment by chai86** [comment]
Hi @elorzafe. I've successfully been able to add new entries and update information using AppSync. However i'm totally stumped when it comes to 'Get' information.

So for example if i try:
` await API.graphql(graphqlOperation(getVideo, { id: ???? }))`  

I have no idea what my 'id' is. Obviously once the entry is created a unique id is created, but i have no way of knowing that unless going to my AWS console. 

I have created my own (as it were in dynamodb, SortKey) called 'videoIntId' in creating the entry which i know (see code from Jan 19). I need to use this videoIntId to retrieve the information in my table. How do i go about using this to get my table entry information?
